충남대 BE_배민수 2주차 과제 (2단계 ~ 3단계)

minsu-cnu opened the pull request:
이번에도 postman으로 대부분의 테스트를 진행했습니다. 추후 mock에 대해 공부하고 테스트 코드를 작성하고자 계획하고 있습니다!

과제가 금요일 18시까지 제출인데, 시간 상 2단계, 3단계의 리뷰 및 반영 과정을 따로 거치기에는 촉박하다고 판단되어서 2단계와 3단계를 하나의 PR로 드리게 되었습니다. 이 점 멘토님께 양해 말씀 드립니다..!

2단계, 3단계를 진행하면서 궁금한 점이 몇 가지 있는데 질문드리고 싶습니다!

<br/>
<hr/>
<br/>


![erd](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/assets/104720635/fa2838d7-3a3b-4a7c-a1ee-1081020479ac)

1. ERD를 설계하면서, 회원 테이블에서 이메일은 일반적으로 변경될 수도 있다는 생각이 들어서, pk로 두기 애매한 것 같아 인조키를 pk로 두었습니다.

    그런데 이런 방식말고 그냥 이메일을 pk로 두어도 상관이 없을지 궁금합니다. 또한 이런 설계가 추후 어떤 문제점이 발생할 여지가 있을지 멘토님의 의견이 궁금합니다!

    또한 위시 리스트 요구 사항을 기존의 상품 테이블을 활용해서 구현할 수 있을 것 같아서, 설계를 약간 바꾸어 위와 같이 상품 테이블을 구성하였는데, 기능은 전체적으로 알맞게 잘 동작하지만.. 제가 요구 사항을 제대로 분석한 것인지 잘 모르겠습니다. 

<br/>
<br/>

2. 이건 질문은 아니지만... 개인적으로 이번 과제가 꽤나 어려웠어서 객체 지향적인 구현과 코드 효율성을 고려하지 못하고, 과제 마감 기한에 맞추기 위해 급하게 구현했다보니.. 메소드 분리같은 것도 잘 안되어있고 코드가 많이 지저분합니다.. 그래서 좀 더 깔끔한 코드를 위해 주말동안 리팩토링을 해볼 계획입니다!

minsu-cnu committed:
refactor: 카카오 문구 관련 유효성 검증 로직을 service layer에서 dto로 옮김

- 관리자 유저는 일반 유저와 달리 카카오 문구 관련 유효성 검증
  로직을 수행할 필요가 없음
- 이로 인해 dto 코드가 달라지므로, dto 추상화 전략을 적용하였음
- service layer에서는 인터페이스형 dto를 받도록하여 코드 변경을 최소화
- ProductController에서는 body를 ClientProductDto에 바인딩
- AdminController에서는 body를 AdminProductDto에 바인딩"

minsu-cnu committed:
refactor: schema.sql 회원 테이블 추가, 상품 테이블 리팩토링

minsu-cnu committed:
docs: step2, step3 기능 명세 작성

minsu-cnu committed:
fix: 콤마 누락, DDL 순서 변경

minsu-cnu committed:
feat: 회원가입 기능 구현

- /members/register 로 들어오는 POST 요청을 처리하는 컨트롤러 구현
- 회원가입 로직을 수행하는 AuthService와 AuthRepository 구현
- 비즈니스적으로 이메일로 회원을 식별함. 중복 시 회원가입 실패

minsu-cnu committed:
chore: jwt 라이브러리 추가

minsu-cnu committed:
feat: 로그인 기능 구현

- 이메일로 회원을 식별하며, 존재하지 않으면 로그인 실패
- 성공 시 응답으로 액세스 토큰 발급

minsu-cnu committed:
fix: 로그인 기능에 비밀번호 검증 로직 추가

minsu-cnu committed:
refactor: 로그인 실패 예외 처리 시 상태코드 403을 응답하도록 변경

minsu-cnu committed:
feat: jwt 생성 및 인터셉터 검증 로직 구현

- application.properties에서 상수로 쓸 값 지정
- 인터셉터의 preHandle에서 jwt 검증 로직 수행
- 검증 성공 시 컨트롤러 메소드에게 email 값 전달

minsu-cnu committed:
refactor: 로그인 시 액세스 토큰에 회원 id 정보도 담도록 변경

minsu-cnu committed:
refactor: 기존 Product CRUD 로직 리팩토링

- 이제는 사용자 별로 각각 독립적으로 위시리스트를 가지므로, CRUD 로직에
  로그인 한 회원의 정보를 활용하도록 리팩토링하였음.
- jwt 액세스 토큰을 생성할 때 회원 id도 담도록 변경
- Product 도메인과 연관된 각 layer에 LoginMember DTO를 추가함.
이를 통해 repository 단에서 DB에 접근할 때 해당 회원의 데이터만 조회 및
조작 가능

minsu-cnu committed:
docs: 기능 명세 수정

minsu-cnu committed:
feat: 관리자 페이지 로그인 기능 구현

- schema.sql에서 관리자 계정 추가하는 쿼리 추가
- Configuration에서 빈에 등록한 토큰검증인터셉터의 path pattern에 '/admin/**' 추가
- 관리자 페이지 로그인 폼 페이지 추가
- 관리자 페이지에 한하여, 클라이언트(타임리프 관리자 페이지)로부터 받은 액세스 토큰을 첫 번째 인터셉터에게 쿠키에 담아 넘겨서, 이걸 풀어서 두번째 토큰 검증 인터셉터에게 넘겨주고 여기서 검증 수행 (일반 사용자의 경우 헤더에 토큰 값이 담겨 있을텐데 이걸 사용)
- 관리자 페이지 로그인 버튼 -> 토큰 값을 RedirectAttributes에 담아서 login process 페이지로 리다이렉트 -> 해당 토큰 값을 hidden input에 담고 자바스크립트로 이걸 post로 submit해서, 컨트롤러의 해당 메소드에서 관리자 위시 리스트 페이지로 리다이렉트하는데, 하기 전에 인터셉터 실행해서 토큰 검증 수행

minsu-cnu committed:
docs: 기능 명세 수정

minsu-cnu committed:
refactor: 관리자 페이지 CRUD 요청 핸들러에 액세스 토큰 기반 회원 정보 활용 로직 추가

- 관리자 페이지는 바닐라 자바스크립트로 헤더에 토큰을 매번 담아보내기 번거로워서, 간단하게 쿠키로 서버에 토큰을 보냄
- 관리자 계정으로 로그인할 때, 성공 시 쿠키에 액세스 토큰 값을 저장하고 이를 관리자 페이지에서 API 요청 할때마다 서버에 보냄

jjongwa commented on src/main/java/gift/product/controller/AdminController.java:
```diff
@@ -1,65 +1,128 @@
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
 
 @Controller
-@RequestMapping("/admin/products")
+@RequestMapping("/admin")
```
여기는 `@Controller`를 사용헀네요!
`@RestController`와 `@Controller`의 차이점은 무엇일까요?

minsu-cnu replied:
`@Controller`는 메소드가 뷰를 반환할 때 사용하는 어노테이션으로 알고있습니다.
`@RestController`는 `@Controller`와 `@ResponseBody`의 조합으로, 모든 메소드가 뷰가 아닌 어떤 객체를 반환해야할 때 사용하는 것으로 알고 있습니다.

원래는 `@Controller`에다가 필요한 메소드에만 `@ResponseBody`를 붙여서 사용하다가, RESTful API에 해당하는 컨트롤러에는 편리하게 활용할 수 있게끔 하기 위해 `@RestController`가 나온 것으로 알고 있습니다.

jjongwa commented on src/main/java/gift/product/controller/AdminController.java:
```diff
@@ -1,65 +1,128 @@
+    }
+
+
+    @GetMapping("/products")
+    public String products(HttpServletRequest request, Model model) {
+        LoginMember loginMember = new LoginMember((Long)request.getAttribute("memberId"),
+            (String)request.getAttribute("email"));
+        List<Product> products = productService.getProductAll(loginMember);
         model.addAttribute("products", products);
         return "admin/products";
     }
```
email 정보는 쓰이지 않는 것 같은데 LoginMember 객체를 만들어 repository까지 전달하는 이유가 있나요?
LoginMember는 어떤 역할을 의도한건지 궁금하네요.!

minsu-cnu replied:
LoginMember는 사용자가 서버에 어떤 자원에 대한 DB CRUD를 요청했을 때, 해당 요청을 보낸 회원의 정보를 레포지토리 계층까지 보냄으로써, 다른 회원 것은 건드리지 않고 해당 회원에 대한 위시리스트 상품만을 조작하게 하기 위한 dto입니다.

그런데 조언 주신 것을 보고 확인해보았는데, 현재 비즈니스 로직에서는 이메일이 중복되는 경우가 없다고 가정했기 때문에 회원을 식별하는데에는 id도 가능하고 이메일도 가능한 상황입니다.

그래서 id나 이메일 둘 중 하나만 dto로 보내도 되고, 말씀 주신 대로 성능 측면에서 좀 더 나은 id를 보내는게 맞는 것 같습니다. 다행히 현재 로직은 id를 보내고 있는데, 문제는 불필요한 이메일 값도 같이 보내고 있습니다. 이 부분은 리뷰 반영하여 제거하겠습니다!

jjongwa commented on src/main/java/gift/product/controller/AdminController.java:
```diff
@@ -1,65 +1,128 @@
+
+    @PostMapping("login/process")
+    @ResponseBody
+    public ResponseEntity<?> loginSuccess(@RequestParam("accessToken") JwtResponse jwtResponse, HttpServletResponse response) {
```
와일드카드 타입을 넣는 것보단 Void를 넣는 방법도 있습니다~!
```suggestion
    public ResponseEntity<Void> loginSuccess(@RequestParam("accessToken") JwtResponse jwtResponse, HttpServletResponse response) {
```

minsu-cnu replied:
조언 감사합니다! 반영하겠습니다.

jjongwa commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -26,34 +28,44 @@ public ProductController(ProductService productService) {
     @DeleteMapping("/delete/{id}")
-    public ResponseEntity<Void> deleteProduct(@PathVariable(name = "id") Long id) {
-        productService.deleteProduct(id);
+    public ResponseEntity<Void> deleteProduct(@PathVariable(name = "id") Long id, HttpServletRequest request) {
+        LoginMember loginMember = new LoginMember((Long)request.getAttribute("memberId"),
+            (String)request.getAttribute("email"));
+        productService.deleteProduct(id, loginMember);
 
         return ResponseEntity.ok().build();
```
지금 현재 구성으로 보았을 때
input으로 받은 id에 해당하는 상품이 없거나,
해당하는 상품이 있는데 로그인한 사람의 상품이 아닐 때에도 동일하게 OK 응답을 보낼 것 같네요.! 
각 상황별 예외 처리를 추가해 보는건 어떨까요?

minsu-cnu replied:
input으로 받은 id에 해당하는 상품이 없는 경우, 또는 해당하는 상품이 존재하지만 로그인 한 사람의 상품이 아닌 경우 모두 레포지토리의 findById를 활용하여 유효성 검증을 할 수 있다고 판단하여, 해당 검증 및 예외 처리를 조회, 수정, 삭제 기능에 대해 추가해보았습니다.

예외 발생 시, 특정 상품이 다른 사용자의 위시리스트에 있다는 것을 숨기기 위해 상태코드를 404로 하였습니다. 또한 같은 이유로 상태 메시지도, 상품이 있지만 다른 사용자의 것일 경우에 똑같이 상품이 없다는 에러 메시지를 응답하도록 구현했습니다.

jjongwa commented on src/main/java/gift/product/dto/AdminProductDto.java:
```diff
@@ -0,0 +1,13 @@
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Size;
+
+public record AdminProductDto(
+    @Size(max = 15, message = "상품 이름은 공백 포함 최대 15자까지 입력할 수 있습니다.")
+    @Pattern(regexp = "^[a-zA-Z0-9가-힣 ()\\[\\]+\\-&/_]*$", message = "사용 가능한 특수 문자는 ()[]+-&/_ 입니다.")
+    String name,
+    int price,
+    String imageUrl) implements ProductDto {
+
+}
```
목적에 따른 Dto 분리 👍

minsu-cnu replied:
감사합니다!

jjongwa commented on src/main/java/gift/product/model/Member.java:
```diff
@@ -0,0 +1,34 @@
+package gift.product.model;
+
+public class Member {
+    private Long memberId;
+    private String email;
+    private String password;
+
+    public Member() {
+
+    }
+
+    public Member(String email, String password) {
+        this.email = email;
+        this.password = password;
+    }
+
+    public Member(Long memberId, String email, String password) {
+        this.memberId = memberId;
+        this.email = email;
+        this.password = password;
+    }
+
+    public Long getMemberId() {
+        return memberId;
+    }
+
+    public String getEmail() {
+        return email;
+    }
+
+    public String getPassword() {
+        return password;
+    }
+}
```
```
public class Member {
    private final Long memberId;
    private final String email;
    private final String password;

    public Member(String email, String password) {
        this.memberId = null;
        this.email = email;
        this.password = password;
    }

    public Member(Long memberId, String email, String password) {
        this.memberId = memberId;
        this.email = email;
        this.password = password;
    }

    public Long getMemberId() {
        return memberId;
    }

    public String getEmail() {
        return email;
    }

    public String getPassword() {
        return password;
    }
}
```

model 패키지에 있는 친구들은 이렇게 필드를 final를 붙여본 다음 전체적인 로직이 동작하게 리팩토링 해보면 조금 더 객체지향적인 코드를 만들 수 있을 거에요! 필수까진 아니지만 추천드립니다~!

minsu-cnu replied:
해당 부분 반영하고, 추가로 생성자 체이닝을 적용해보았습니다!

jjongwa replied:
오 아주 좋습니다! 👍👍

jjongwa commented on src/main/java/gift/product/repository/AuthRepository.java:
```diff
@@ -0,0 +1,37 @@
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    public boolean existsByEmail(String email) {
+        var sql = "SELECT * FROM Member WHERE email = ?";
+        boolean isMemberNotExist = jdbcTemplate.query(sql, (resultSet, rowNum) -> 0, email).isEmpty();
+
+        return !isMemberNotExist;
+    }
```
부정의 부정을 쓰다 보면 헷갈릴 수 있어요! 리팩토링 해볼까요?

minsu-cnu replied:
해당 부분 반영하여, 긍정으로 변경하였는데 변수 할당 후 단순 리턴하고 있는 구조가 되어서, 리턴문에 표현식을 모두 담도록 리팩토링을 하였습니다.

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -20,19 +22,20 @@ public ProductRepository(JdbcTemplate jdbcTemplate) {
         KeyHolder keyHolder = new GeneratedKeyHolder();
-        var sql = "INSERT INTO Product (name, price, imageUrl) VALUES (?, ?, ?)";
+        var sql = "INSERT INTO Product (member_id, name, price, imageUrl) VALUES (?, ?, ?, ?)";
 
         jdbcTemplate.update(new PreparedStatementCreator() {
             @Override
             public PreparedStatement createPreparedStatement(Connection con) throws SQLException {
                 PreparedStatement pstmt = con.prepareStatement(
                     sql, new String[]{"id"}
                 );
-                pstmt.setString(1, product.getName());
-                pstmt.setInt(2, product.getPrice());
-                pstmt.setString(3, product.getImageUrl());
+                pstmt.setLong(1, loginMember.memberId());
+                pstmt.setString(2, product.getName());
+                pstmt.setInt(3, product.getPrice());
+                pstmt.setString(4, product.getImageUrl());
 
                 return pstmt;
             }
```
`SimpleJdbcInsert`도 한번 학습해 보고 사용해 볼까요?

minsu-cnu replied:
말씀해주신 SimpleJdbcInsert를 활용하여, 기존에 KeyHolder와 PreparedStatementCreator를 활용하던 복잡한 save 메소드 로직을 간단한 구조로 리팩토링할 수 있었습니다. 또한 AuthRepository의 registerMember도 SimpleJdbcInsert를 활용하여 간단하게 리팩토링해보았습니다.

jjongwa reviewed (changes requested):
2, 3단계도 잘 구현해 주셨어요~! 👏👏

주말동안 리펙토링과 테스트 부분 보강한다고 하셔서 간단한 리뷰 정도 남겼습니다!

질문 주신 부분에 대한 답을 하자면..
Member의 PK를 회원 고유번호로 지정한건 아주 잘 선택해 주셨어요!
민수님이 말씀하신대로 이메일은 추후 변경될 여지가 있을 수도 있구요,
PK는 결국 각 데이터를 **고유하게 분별**할 수 있고, 그러면서 **빠르게** 찾을 수 있어야 하는게 포인트입니다!
만약 이메일같은 긴 문자열로 키를 설정한다면 조회할 때 상대적으로 시간이 많이 걸리겠죠?
DB의 인덱스에 대해서 공부해 보시면 바로 감이 올거예요!

주말동안 리팩토링 진행하면서 추가적으로 궁금한게 생기면 뭐든 질문 남겨주세요~!

minsu-cnu committed:
refactor: 코드 리뷰 반영

- LoginMember DTO가 기존의 member_id, email을 운반하던 것에서
  member_id만 담도록 변경
- AdminController의 loginSuccess 메소드에서 와일드카드 제네릭 타입을
  Void로 변경
- 존재하지 않는 상품에 대한 요청 또는 다른 사용자의 상품에 대한 요청에
  대한 예외 처리 추가
- model에 있는 도메인들의 필드를 final로 바꾸고, 생성자 체이닝을 적용
- AuthRepository의 existsByEmail에서 부정의 부정을 긍정으로 변경
- repository의 모든 DB INSERT 관련 메소드를 SimpleJdbcInsert를
  활용하도록 리팩토링

minsu-cnu committed:
refactor: 전체 코드 메소드 분리 적용

- AdminController 메소드 분리 (getLoginMember, addAccessTokenCookieInResponse)
- ProductController 메소드 분리 (getLoginMember)
- ProductRepository 메소드 분리 (getProductRowMapper)
- AuthService 메소드 분리 (getAccessToken)
- GlobalExceptionHandler 메소드 분리 (buildErrorMessage)
- JwtCookieToHeaderInterceptor 리팩토링 (메소드 분리, null 체크 로직 변경, 에러 메시지 변경)
- TokenValidationInterceptor 메소드 분리 (decodeAccessToken, getAccessTokenFromCustomHeader, getAccessTokenFromHeader)

minsu-cnu committed:
test: 테스트 코드 리팩토링 및 추가

- ClientProductTest 테스트 메소드명 한글로 변경, service의 메소드에 필요한 LoginMember DTO 추가
- 회원 가입 및 로그인을 테스트하는 클래스 추가

minsu-cnu committed:
style: 자바 코드 컨벤션 적용

minsu-cnu committed:
fix: 토큰 검증 인터셉터의 대상 경로에서 관리자 페이지 로그인 경로 제외

jjongwa reviewed (approved):
리뷰 반영 잘 해주셨어요!  
지난 1주차보다 고민의 흔적이 물씬 드러나는 코드였습니다~! 💯 

이번 주차 미션은 이만 머지하겠습니다! 남은 주말 푹 쉬고 3주차에서 만나요~!
